### PR TITLE
Enhancement/scraper vine

### PR DESCRIPTION
**Pull request by ayyubibrahimi**
What do you think of this version of the booking data scraper? My plan is to scrape the booking rosters once a week.

### CODE REVIEW

File: extraction/scraper_vine.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    chrome_options.add_argument('--disable-gpu')
+    chrome_options.add_argument('--log-level=3')
+    browser = webdriver.Chrome(chrome_options=chrome_options)
+    browser.get(url)
+    login_button = WebDriverWait(browser, 2).until(EC.presence_of_element_located((By.ID, 'lbShowAll')))
+    login_button.click()
+
+    try:
+        inmates = []
+        soup = BeautifulSoup(browser.page_source, 'lxml')
```
Comment (pckhoi) [inline extraction/scraper_vine.py:28]: If you are only interested in the page source then I think it's better to not use Selenium. Selenium is clunky to install cross-OS and is super slow.

Also, I recommend using [scrapy](https://docs.scrapy.org/en/latest/) for anything crawling-related.

Comment (pckhoi) [reply extraction/scraper_vine.py:28]: Also committing a 10 MiB binary file like Chrome.exe is a big no-no. The common way to treat binary files is to only provide a download link or a script to install but not the file itself. Also this binary file doesn't work outside of Windows. But then again, using Selenium just to get page source is like skinning a chicken with a chainsaw.

Comment (ayyubibrahimi) [reply extraction/scraper_vine.py:28]: The html for the roster tables is hidden for a majority of the jails. This scraping method was the only one that proved successful after trying a few different approaches, spacy methods included. I'll give it another try.

File: extraction/scraper_vine.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    oakdale = 'http://www.oakdale.lavns.org/roster.aspx'
+    oakdale = scrape(oakdale).to_csv(data_file_path('scraped/vine/oakdale_pd_booking.csv'), index=False)
+    shreveport = 'http://www.shreveport.lavns.org/roster.aspx'
+    shreveport = scrape(shreveport).to_csv(data_file_path('scraped/vine/shreveport_pd_booking.csv'), index=False)
+    sulphur = 'http://www.sulphur.lavns.org/roster.aspx'
+    sulphur = scrape(sulphur).to_csv(data_file_path('scraped/vine/sulphur_pd_booking.csv'), index=False)
+    villeplatte = 'http://www.villeplatte.lavns.org/roster.aspx'
+    villeplatte = scrape(villeplatte).to_csv(data_file_path('scraped/vine/villeplatte_pd_booking.csv'), index=False)
+    winnfield = 'http://www.winnfield.lavns.org/roster.aspx'
+    winnfield = scrape(winnfield).to_csv(data_file_path('scraped/vine/winnfield_pd_booking.csv'), index=False)
```
Comment (pckhoi) [inline extraction/scraper_vine.py:188]: It seems that this part could use a mapping between agency name and URL to scrape. If you're not afraid of diving into `make` config, you can extract this mapping into a json file (i.e. `raw_datasets.json`) for ease of maintenance. But before worrying about it, I recommend replacing Selenium+BeautifulSoup with Scrapy first.

**Comment by pckhoi** [comment]
And why does this branch also contain changes for Lafayette?

File: extraction/scraper_vine.py (hunk 2 above)
Comment (ayyubibrahimi) [reply extraction/scraper_vine.py:188]: I'll look into it. Thanks.

**Comment by ayyubibrahimi** [comment]
Not sure at the moment. I'm certain that I added commit(s) and pushed to the Lafayette branch before working on scripts from this branch.

File: extraction/scraper_vine.py (hunk 1 above)
Comment (pckhoi) [reply extraction/scraper_vine.py:28]: I think you'll have success with Scrapy. The data does seem to be present in page source.

Comment (pckhoi) [reply extraction/scraper_vine.py:28]: Which jails have hidden roster tables?

Comment (ayyubibrahimi) [reply extraction/scraper_vine.py:28]: I also noticed that it is present, which is why I looked for alternative solutions when I received the following output from Scrapy: 

Beauregard Parish Sheriff's Office
 Jail Roster
                                Click to register anonymously to be notified
                                upon any changes in this offender's custody status.
                                             Please wait...
Filter:
Show All
A


B

C

D

E

F

G

H

I

J   

K

L

M

N

O

P

Q

R

S

T

U

V

W

X

Y

Z
 # of offenders: Last updated: 11/19/2021 8:31:13 PM Central Time
Beauregard Parish Sheriff's Office
 Jail Roster
Click to register anonymously to be notified
                                upon any changes in this offender's custody status
                                             Please wait...
Filter:
Show All


A


B

C

D

E

F

G

H

I

J   

K

L

M

N

O

P

Q

R

S

T

U

V

W

X

Y

Z
 # of offenders: Last updated: 11/19/2021 8:31:13 PM Central Time
Please wait....
Filter:
Show All
A
B
C
D
E
F
G
H
I
J
K
L
M
N
O
P
Q
R
S
T
U
V
W
X
Y
Z
# of offenders: Last updated: 11/19/2021 8:31:13 PM Central Time


This agency is providing this roster of incarcerated offenders to the public and law enforcement in the interest of public safety. This information shall not be considered, or used as, a public document, or official document, and no other publication or copying of this information is allowed without the express written consent of this agency. Any unauthorized use of this information is forbidden and subject to criminal prosecution.

It's only when I use Selenium and Google Drive, which works to stagger to request/replicate a user, that I receive the intended output.

File: extraction/scraper_vine.py (hunk 2 above)
Comment (pckhoi) [reply extraction/scraper_vine.py:188]: Actually, don't worry about make config. Now that I think more about it, putting the links in a simple dictionary is good enough.

File: extraction/scraper_vine.py (hunk 1 above)
Comment (ayyubibrahimi) [reply extraction/scraper_vine.py:28]: I should make clear that I am able to scrape the rosters with scrapy for 4/54 agencies. I only looked for alternative methods after receiving the above output for the remaining agencies.

Comment (pckhoi) [reply extraction/scraper_vine.py:28]: I can take a look at the agencies that are not working if you have already committed the code.

Comment (ayyubibrahimi) [reply extraction/scraper_vine.py:28]: I haven't yet. I'll commit the scrapy code soon for your review. Thanks.

**Comment by ayyubibrahimi** [comment]
```
import re

from scrapy.http import FormRequest
from scrapy.spiders import Spider


HEADERS = {
    'X-MicrosoftAjax': 'Delta=true',
    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.76 Safari/537.36'
}


URL = 'http://www.jefferson.lavns.org/'


class ExitRealtySpider(Spider):
    name = "jpso"

    allowed_domains = ["jefferson.lavns.org"]
    start_urls = [URL]


    def parse(self, response):
        # submit a form (first page)
        self.data = {}
        for post in response.css('tr'):
            yield {
                'name': post.css('td')[0].get(),
            }

        self.data['ctl00$MainContent$ScriptManager1'] = 'ctl00$MainContent$UpdatePanel1|ctl00$MainContent$agentList'
        self.data['__EVENTTARGET'] = 'ctl00$MainContent$List'
        self.data['__EVENTARGUMENT'] = 'Page$1'

        return FormRequest(url=URL,
                           method='POST',
                           callback=self.parse_page,
                           formdata=self.data,
                           meta={'page': 1},
                           dont_filter=True,
                           headers=HEADERS)


    def parse_page(self, response):
        current_page = response.meta['page'] + 1

        # parse agents (TODO: yield items instead of printing)
        for agent in response.xpath('tr'):
            print (agent.extract())
        print ("------")

        # request the next page
        data = {
            '__EVENTARGUMENT': 'Page$%d' % current_page,
            '__EVENTVALIDATION': re.search(r"__EVENTVALIDATION\|(.*?)\|", response.body, re.MULTILINE).group(1),
            '__VIEWSTATE': re.search(r"__VIEWSTATE\|(.*?)\|", response.body, re.MULTILINE).group(1),
            '__ASYNCPOST': 'true',
            '__EVENTTARGET': 'ctl00$MainContent$agentList',
            'ctl00$MainContent$ScriptManager1': 'ctl00$MainContent$UpdatePanel1|ctl00$MainContent$agentList',
            '': ''
        }

        return FormRequest(url=URL,
                           method='POST',
                           formdata=data,
                           callback=self.parse_page,
                           meta={'page': current_page},
                           dont_filter=True,
                           headers=HEADERS)

```

Here is the scrapy code which returns the message: "

“This agency is providing this roster of incarcerated offenders to the public and law enforcement in the interest of public safety. This information shall not be considered, or used as, a public document, or official document, and no other publication or copying of this information is allowed without the express written consent of this agency. Any unauthorized use of this information is forbidden and subject to criminal prosecution.” 

I received the same message, rather than roster data, when I used scrapy, bs4 and selenium. After a bit more experimentation, only the code that I pushed, which combines both bs4 and selenium, returns the roster data. I'll keep experimenting.

**Comment by pckhoi** [comment]
Try this:
```python
class ExitRealtySpider(Spider):
    name = "jpso"

    allowed_domains = ["jefferson.lavns.org"]

    def start_requests(self):
        urls = [URL]
        for url in urls:
            yield FormRequest(
                url=url,
                method='GET',
                callback=self.parse,
                formdata={
                    '__EVENTTARGET': 'lbShowAll'
                }
            )

    def parse(self, response):
        # submit a form (first page)
        # self.data = {}
        for tr in response.css('table#gvRoster tr')[1:]:
            yield {
                'name': tr.css('td:nth-child(2)::text')[0].get(),
            }
```

Also, a single spider should be able to handle all links from all jails, so let's change the name from "jpso" to something else.

**Comment by ayyubibrahimi** [comment]
Amazing. Works perfectly. Thanks!

**Comment by ayyubibrahimi** [comment]
I will have this ready for review soon.

File: fuse/lafayette_pd.py [hunk 3]
```diff
@@ -43,24 +43,33 @@ def fuse_events(cprr, pprr):
...
-        cprr[['uid', 'first_name', 'last_name']],
-        cprr[['investigator_uid', 'investigator_first_name', 'investigator_last_name']].rename(columns={
+        cprr_20[['uid', 'first_name', 'last_name']],
+        cprr_20[['investigator_uid', 'investigator_first_name', 'investigator_last_name']].rename(columns={
+            'investigator_uid': 'uid',
+            'investigator_first_name': 'first_name',
+            'investigator_last_name': 'last_name',
+        }),
+        cprr_14[['uid', 'first_name', 'last_name']],
+        cprr_14[['investigator_uid', 'investigator_first_name', 'investigator_last_name']].rename(columns={
```
Comment (pckhoi) [inline fuse/lafayette_pd.py:72]: What with the lafayette changes?

File: lib/clean.py [hunk 4]
```diff
@@ -567,3 +567,34 @@ def remove_future_dates(df: pd.DataFrame, max_date: str, prefixes: List[str]) ->
...
+            the frame to process
+        cols (list of str):
+            birth date columns
+
+    Returns:
+        the updated frame
+    """
+    for col in cols:
+        df.loc[:, col] = strip_birth_day_and_month(df[col])
+    return df
```
Comment (pckhoi) [inline lib/clean.py:600]: I don't see these functions being used anywhere.

**Comment by pckhoi** [comment]
Very nice! But please put unrelated changes into a separate PR. Also, where are the scraped outputs saved?

File: fuse/lafayette_pd.py (hunk 3 above)
Comment (ayyubibrahimi) [reply fuse/lafayette_pd.py:72]: They are a product of the merge. I notice errors. I'll fix them in the bugs/misc branch.

File: lib/clean.py (hunk 4 above)
Comment (ayyubibrahimi) [reply lib/clean.py:600]: These functions will be used when I clean the booking data.

I figure that I should do the cleaning of these data in a separate script. Do you agree?

**Comment by ayyubibrahimi** [comment]
For now, they are saved in the same directory as the scraper. My plan was to upload it to Dropbox, put that new link into `raw_datasets.json,` and proceed to clean from that step. Thoughts?

File: lib/clean.py (hunk 4 above)
Comment (pckhoi) [reply lib/clean.py:600]: If they are reuseable then it's fine to leave them here.

**Comment by pckhoi** [comment]
Yup, sounds good to me. Putting them on Wrgl is another option but I'm still figuring out the sync replacement atm.

**Review by pckhoi — APPROVED** [review]
Everything looks fine but be careful with the Lafayette changes.

**Pull request merged**